File: ScheduleKitTests/SCKDayPointTests.m
/*
 *  SCKDayPointTests.h
 *  ScheduleKitTests
 *
 *  Created:    Guillem Servera on 05/02/2015.
 *  Copyright:  © 2014-2015 Guillem Servera (http://github.com/gservera)
 *
 *  Permission is hereby granted, free of charge, to any person obtaining a copy
 *  of this software and associated documentation files (the "Software"), to deal
 *  in the Software without restriction, including without limitation the rights
 *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *  copies of the Software, and to permit persons to whom the Software is
 *  furnished to do so, subject to the following conditions:
 *
 *  The above copyright notice and this permission notice shall be included in
 *  all copies or substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *  THE SOFTWARE.
 */

@import Cocoa;
@import XCTest;

#import "SCKDayPoint.h"

@interface SCKDayPointTests : XCTestCase
@end

@implementation SCKDayPointTests

- (void)testDesignatedInit {
    SCKDayPoint *p1 = [[SCKDayPoint alloc] initWithHour:18 minute:30 second:25];
    XCTAssertNotNil(p1,@"SCKDayPoint initialization failed");
    XCTAssertEqual(p1.hour, 18, @"SCKDayPoint initialization failed");
    XCTAssertEqual(p1.minute, 30, @"SCKDayPoint initialization failed");
    XCTAssertEqual(p1.second, 25, @"SCKDayPoint initialization failed");
    
    SCKDayPoint *p2 = [[SCKDayPoint alloc] initWithHour:18 minute:30 second:65];
    XCTAssertNotNil(p2,@"SCKDayPoint initialization failed");
    XCTAssertEqual(p2.hour, 18, @"SCKDayPoint initialization failed");
    XCTAssertEqual(p2.minute, 31, @"SCKDayPoint initialization failed");
    XCTAssertEqual(p2.second, 5, @"SCKDayPoint initialization failed");
    
    SCKDayPoint *p3 = [[SCKDayPoint alloc] initWithHour:18 minute:65 second:65];
    XCTAssertNotNil(p3,@"SCKDayPoint initialization failed");
    XCTAssertEqual(p3.hour, 19, @"SCKDayPoint initialization failed");
    XCTAssertEqual(p3.minute, 6, @"SCKDayPoint initialization failed");
    XCTAssertEqual(p3.second, 5, @"SCKDayPoint initialization failed");
    
    SCKDayPoint *p4 = [[SCKDayPoint alloc] initWithHour:0 minute:0 second:-65];
    XCTAssertNotNil(p4,@"SCKDayPoint initialization failed");
    XCTAssertEqual(p4.hour, 0, @"SCKDayPoint initialization failed");
    XCTAssertEqual(p4.minute, -1, @"SCKDayPoint initialization failed");
    XCTAssertEqual(p4.second, -5, @"SCKDayPoint initialization failed");
    
    SCKDayPoint *p5 = [[SCKDayPoint alloc] initWithHour:12 minute:-60 second:-3600];
    XCTAssertNotNil(p5,@"SCKDayPoint initialization failed");
    XCTAssertEqual(p5.hour, 10, @"SCKDayPoint initialization failed");
    XCTAssertEqual(p5.minute, 0, @"SCKDayPoint initialization failed");
    XCTAssertEqual(p5.second, 0, @"SCKDayPoint initialization failed");
}

- (void)testZeroInit {
    SCKDayPoint *zP = [SCKDayPoint zeroPoint];
    XCTAssertNotNil(zP,@"SCKDayPoint zero initialization failed");
    XCTAssertEqual(zP.hour, 0, @"SCKDayPoint zero initialization failed");
    XCTAssertEqual(zP.minute, 0, @"SCKDayPoint zero initialization failed");
    XCTAssertEqual(zP.second, 0, @"SCKDayPoint zero initialization failed");
}

- (void)testDateInit {
    NSDate *d = [[NSCalendar currentCalendar] dateBySettingHour:18 minute:30 second:25 ofDate:[NSDate date] options:0];
    SCKDayPoint *zP = [[SCKDayPoint alloc] initWithDate:d];
    XCTAssertNotNil(zP,@"SCKDayPoint date initialization failed");
    XCTAssertEqual(zP.hour, 18, @"SCKDayPoint date initialization failed");
    XCTAssertEqual(zP.minute, 30, @"SCKDayPoint date initialization failed");
    XCTAssertEqual(zP.second, 25, @"SCKDayPoint date initialization failed");
}

- (void)testDayOffset {
    SCKDayPoint *p1 = [[SCKDayPoint alloc] initWithHour:18 minute:30 second:25];
    XCTAssertEqual(p1.dayOffset, 66625, @"Day offset calculation failed");
}

- (void)testComparsion {
    SCKDayPoint *p1 = [[SCKDayPoint alloc] initWithHour:1 minute:50 second:55];
    SCKDayPoint *p2 = [[SCKDayPoint alloc] initWithHour:2 minute:35 second:1];
    SCKDayPoint *p3 = [[SCKDayPoint alloc] initWithHour:2 minute:35 second:1];
    SCKDayPoint *p4 = [[SCKDayPoint alloc] initWithHour:3 minute:17 second:56];
    XCTAssertTrue([p2 isEqual:p3],@"SCKDayPoint comparsion failed");
    XCTAssertFalse([p2 isEqual:p1],@"SCKDayPoint comparsion failed");
    XCTAssertTrue([p2 isEqualToDayPoint:p3],@"SCKDayPoint comparsion failed");
    XCTAssertFalse([p2 isEqualToDayPoint:p1],@"SCKDayPoint comparsion failed");
    XCTAssertTrue([p2 isLaterThanDayPoint:p1],@"SCKDayPoint comparsion failed");
    XCTAssertFalse([p2 isLaterThanDayPoint:p4],@"SCKDayPoint comparsion failed");
    XCTAssertTrue([p2 isEarlierThanDayPoint:p4],@"SCKDayPoint comparsion failed");
    XCTAssertFalse([p2 isEarlierThanDayPoint:p1],@"SCKDayPoint comparsion failed");
    XCTAssertFalse([p2 isEarlierThanDayPoint:p3],@"SCKDayPoint comparsion failed");
    XCTAssertFalse([p2 isLaterThanDayPoint:p3],@"SCKDayPoint comparsion failed");
}

@end

File: ScheduleKitTests/SCKDayPointTests.swift
//
//  SCKDayPointTests.swift
//  ScheduleKit
//
//  Created by Guillem Servera Negre on 10/12/16.
//  Copyright © 2016 Guillem Servera. All rights reserved.
//

import XCTest
@testable import ScheduleKit

final class SCKDayPointTests: XCTestCase {

    func testInitialization() {
        let point = SCKDayPoint(hour: 10, minute: 20, second: 30)
        XCTAssertEqual(point.hour, 10)
        XCTAssertEqual(point.minute, 20)
        XCTAssertEqual(point.second, 30)
        XCTAssertEqual(point.dayOffset, 37230)

        let calendar = Calendar.current
        let d = calendar.date(bySettingHour: 10, minute: 20, second: 30, of: Date())
        let point2 = SCKDayPoint(date: d!)
        XCTAssertEqual(point2.hour, 10)
        XCTAssertEqual(point2.minute, 20)
        XCTAssertEqual(point2.second, 30)
        XCTAssertEqual(point2.dayOffset, 37230)
    }

    func testInitializationWithOverflow() {
        let point = SCKDayPoint(hour: 0, minute: 0, second: 37230)
        XCTAssertEqual(point.hour, 10)
        XCTAssertEqual(point.minute, 20)
        XCTAssertEqual(point.second, 30)
        XCTAssertEqual(point.dayOffset, 37230)
    }

    func testInitializationWithOverflow2() {
        let point = SCKDayPoint(hour: 10, minute: 21, second: -20)
        XCTAssertEqual(point.hour, 10)
        XCTAssertEqual(point.minute, 20)
        XCTAssertEqual(point.second, 40)
        XCTAssertEqual(point.dayOffset, 37240)

        let point2 = SCKDayPoint(hour: 10, minute: 22, second: -80)
        XCTAssertEqual(point2.hour, 10)
        XCTAssertEqual(point2.minute, 20)
        XCTAssertEqual(point2.second, 40)
        XCTAssertEqual(point2.dayOffset, 37240)
    }

    func testInitializationWithOverflow3() {
        let point = SCKDayPoint(hour: 11, minute: -40, second: 30)
        XCTAssertEqual(point.hour, 10)
        XCTAssertEqual(point.minute, 20)
        XCTAssertEqual(point.second, 30)
        XCTAssertEqual(point.dayOffset, 37230)
        let point2 = SCKDayPoint(hour: 12, minute: -100, second: 30)
        XCTAssertEqual(point2.hour, 10)
        XCTAssertEqual(point2.minute, 20)
        XCTAssertEqual(point2.second, 30)
        XCTAssertEqual(point2.dayOffset, 37230)
    }

    func testComparsion() {
        let point = SCKDayPoint(hour: 9, minute: 30, second: 1)
        let point2 = SCKDayPoint(hour: 10, minute: 29, second: 6)
        let point3 = SCKDayPoint(hour: 9, minute: 30, second: 1)
        XCTAssertEqual(point, point3)
        XCTAssertNotEqual(point, point2)
        XCTAssertTrue(point < point2)
        XCTAssertTrue(point2 > point)
        XCTAssertEqual(point.hashValue, point3.hashValue)
        XCTAssertNotEqual(point.hashValue, point2.hashValue)
    }

}
